### PR TITLE
Add WorkerNetworkBandwidth chart to dashboard

### LINKED ISSUE #5090: Observed worker network bandwidth chart
We have a couple of charts for inter-worker bandwidth that use our own metrics.  I don't trust these metrics that much.  We also keep per-worker metrics.  These are currently reported in the /workers tab with read/write at the end of the table.  I think that these are recorded with psutil and are sent in the worker heartbeats.  We might want to create another real-time bar-chart like occupancy/cpu/memory but for read/write traffic.  

This is slightly more complex because there are a couple of different values to show for each worker.  As options I could imagine ...

1.  Summing these up
2. Overplotting both with colors that make it clear which one is ahead
3. Plotting two skinnier rectangles per worker

### PR DESCRIPTION
**Pull request by ncclementi**
- [x] Closes #5090
- [ ] Tests added / passed
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

Prototype for and observed worker network bandwidth chart. 
You can find the new chart under the "More" tab -> "Network Bandwidth Workers" (we should probably change the name)
Pending:
- Tests 
- Some kind of hover? do we need this or not? 

cc: @mrocklin

### CODE REVIEW

**Comment by mrocklin** [comment]
Screenshot for reference (I'm also including memory for comparison)

image

**Comment by mrocklin** [comment]
This is great!  I ran this on a workload and have a few thoughts:

1.  The x range will need to be responsive I think.  Right now it's set to the worker's memory limit, but that's divorced from the network bandwidth that we might see.  In general network bandwidths will be an order of magnitude less.  I recommend that we dynamically set the maximum xrange to the greater of ...
     - 100 MB/s
     - The maximum value of the current data
     - 0.95 of the previous value

    No hardware is capable of less than 100MB/s these days, so it's a good minimum.  If we go higher than this then great, let's bump up that limit.  The last thing is there to make sure that we don't jump around too much.  This smoothly brings the value down over time, rather than having it snap back to 100MB/s instantly.

We should probably remove the minor tick marks, the horizontal lines, the y-axis, and the Boke logo, as is done in the worker plot (although I think that https://github.com/dask/distributed/pull/5102 should handle the logo for you if you merge main)

**Comment by ncclementi** [comment]
>     1. The x range will need to be responsive I think.  Right now it's set to the worker's memory limit, but that's divorced from the network bandwidth that we might see.  In general network bandwidths will be an order of magnitude less.  I recommend that we dynamically set the maximum xrange to the greater of ...
>        
>        * 100 MB/s
>        * The maximum value of the current data
>        * 0.95 of the previous value
>        
>        No hardware is capable of less than 100MB/s these days, so it's a good minimum.  If we go higher than this then great, let's bump up that limit.  The last thing is there to make sure that we don't jump around too much.  This smoothly brings the value down over time, rather than having it snap back to 100MB/s instantly.

I made an attempt to capture this, I wasn't sure what's the best way to set those 100MB, if you can take a look and comment on the code if there is a better way, happy to change it. 

> We should probably remove the minor tick marks, the horizontal lines, the y-axis, and the Boke logo, as is done in the worker plot (although I think that #5102 should handle the logo for you if you merge main)

I took care of all these, the chart should look cleaner now. 

**Extra:** The name of the plot might no be the best one, do we have a better suggestion for the name of the plot that would be convey better what the plot is?

File: distributed/dashboard/components/scheduler.py
Comment (mrocklin) [inline distributed/dashboard/components/scheduler.py:820]:
```suggestion
            max_limit = max(
                x_read,
                x_write,
                100_000_000,
                self.root.x_range.end,
            )

```

Let's compute this outside of the for loop, and compare against the old version of the x_range.

Comment (mrocklin) [reply distributed/dashboard/components/scheduler.py:820]: The 95% thing meant to drop down a bit from the last cycle of running this function, not the last worker's memory

Comment (ncclementi) [reply distributed/dashboard/components/scheduler.py:820]: Running this snippet around 3 times 

 ```python
 import dask.array as da
x = da.random.random((10000, 10000))
y = (x + x.T).compute()

 ```
Previous version: I see that it resets back to ~100MB

old version

 

With the change you suggest, before doing any computation we have the x_range max to be around 100MiB. Then if I run the computation once I see that the x_range limit goes up to 1.3 GB when it finishes it stays in there. If I run it again, the max will go up to 2 GB and stays there (not sure if this is something we desire) 

matts_version

Comment (ncclementi) [reply distributed/dashboard/components/scheduler.py:820]: For the record haven't pushed any changes with your suggestion yet, I tested this locally, trying to see what behavior do we want to capture.

Comment (mrocklin) [reply distributed/dashboard/components/scheduler.py:820]: It looks like I forgot the `0.95 * ` on the self.root.x_range.end bit.

File: distributed/dashboard/components/scheduler.py
Comment (jrbourbeau) [inline distributed/dashboard/components/scheduler.py:800]: The extra `list` conversion isn't needed since `len` and iteration works on dictionary views

```suggestion
            workers = self.scheduler.workers.values()
```

File: distributed/dashboard/components/scheduler.py
Comment (ncclementi) [reply distributed/dashboard/components/scheduler.py]: I added the `0.95 *`  and that causes that after a computation it'll readjust until the max is 100MB, it'll update few times.  I don't know if this is a problem or not, I don't have a strong opinion on it. 

update


I included in the last push labels to the existing bandwidth time-series under /system to be consistent.

File: distributed/dashboard/components/scheduler.py
Comment (jrbourbeau) [inline distributed/dashboard/components/scheduler.py:803]: Since NumPy may not be installed

https://github.com/dask/distributed/blob/95b83a41badda35111457da1e3e73ae75501fa63/distributed/dashboard/components/scheduler.py#L50-L53

let's use Python's built-in `range` here instead. Something like 

```
y_read = [i + 0.75 for i in range(len(workers))]
```

should be equivalent (though I've not tested the code snippet above)

File: distributed/dashboard/components/scheduler.py
Comment (jrbourbeau) [inline distributed/dashboard/components/scheduler.py:826]: Since we only compute `max_limit` to determine `self.root.x_range.end`, let's have those happen right next to each other. Currently we define `result` between `max_limit` and using `max_limit`. Typically defining variables near where they're used will make it easier for code readers to comprehend. 

Further, since `max_limit` is only used in one place, we might even avoid creating a new variable altogether:

```python
self.root.x_range.end = max(
                max(x_read),
                max(x_write),
                100_000_000,
                0.95 * self.root.x_range.end,
            )
```

File: distributed/dashboard/components/scheduler.py
Comment (ncclementi) [reply distributed/dashboard/components/scheduler.py:800]: Interesting, I noticed that we used it in the Occupancy charts so it carried from a copy and paste. Should we remove the `list()` from the Occupancy chart too then? 

https://github.com/dask/distributed/blob/19cd1f4a6241f2998050bbee9501393797f9ba0a/distributed/dashboard/components/scheduler.py#L145

Comment (jrbourbeau) [reply distributed/dashboard/components/scheduler.py:800]: I see. That's also not needed there too. If you search through that module you'll see we use `workers = self.scheduler.workers.values()` in most places. This is also a small point, but just wanted to point it out for reference

**Comment by ncclementi** [comment]
After a chat with @jrbourbeau  we noticed that having a space between the workers helps to identify where the difference between workers happens. I added such space and it looks better. If we have objections against it, we can always remove it. 

worker_space

**Comment by mrocklin** [comment]
This seems great.  Are we good to go?  If so I suggest that we merge and
move on.

**Comment by ncclementi** [comment]
I'm seeing some tests failing, and it's not quite clear to me why this is happening. 
In one of the ubuntu 3.8 I'm getting
```
assert s.workers[a.address].metrics["read_bytes"] == d["x_read"][0]
E       assert 140757.15019158693 == 145158.18565783507
E         +140757.15019158693
E         -145158.18565783507
```
While the other fails seem to be unrelated. any ideas?

**Comment by jrbourbeau** [comment]
That test failure is triggered by the system monitor periodic callback happening in the background. If an extra system monitor update happens between when we update the `WorkerNetworkBandwidth` source data and the test `assert`s, then we can have a mismatch between the `WorkerNetworkBandwidth` data and the `read_bytes`/`write_bytes` metrics stored on the scheduler. I pushed up a small commit which disables the system monitor periodic callback on the workers to ensure this situation doesn't happen.

**Comment by ncclementi** [comment]
Mmh interesting, although it seems that didn't fix it.

**Comment by mrocklin** [comment]
Checking in here.  Is this still blocked on the issue above?

**Comment by jrbourbeau** [comment]
Yes, but hopefully not for long. I think this might be failing occasionally (I've not been able to reproduce locally) because we're assuming the order of `scheduler.workers` is the same as the order of the worker (e.g. `a` / `b`) generated by `gen_cluster`, which isn't guarenteed to always be `True`. For example, there are [other tests](https://github.com/dask/distributed/blob/bdd46f16ef2bb429e4013c9d3f596421dafac55f/distributed/dashboard/tests/test_scheduler_bokeh.py#L372-L374) where we explicitly don't make that ordering assumption. I'll push up a quick related commit to see if that fixes things

**Comment by jrbourbeau** [comment]
Alright, that should resolve the failure we were seeing. I confirmed locally that the latest commit results in `test_WorkerNetworkBandwidth_metrics` passing, regardless of what order the workers created by `gen_cluster` are in

**Comment by jrbourbeau** [comment]
The tests added here are passing now. @ncclementi could you add a new entry [here](https://github.com/dask/distributed/blob/b37ac9d2032413d745887f1d593a9cdac917e332/docs/source/http_services.rst) for this plot? Otherwise it looks like it's good to merge

**Comment by ncclementi** [comment]
@jrbourbeau you got it! 
I went ahead and also update the endpoint names for the worker memory/cluster plots that were outdated.

**Review by jrbourbeau — APPROVED** [review]
Thanks for working on this @ncclementi! Will merge once CI finishes up

**Comment by mrocklin** [comment]
Woo!

**Pull request merged**
